### DISCUSSION TITLE
Idle recharging,stop charging and execute the new order?

### DISCUSSION THREAD

**Discussion by yuantingxue2028** [openTCS/opentcs#171] [Q&A]
I want to charge when idle, but when there is a new order and the vehicle is charging, how should I stop charging and execute the new order? In the vehicle, I can only determine if the current battery level is greater than the "Sufficiently" energy level, but I cannot determine if a new order has been generated, so I cannot execute the stop charging task. Please guide me on how to handle this. Thank you very much.

**Comment by yuantingxue2028** [comment]
this my code

private void chargeStatusGetPolling(MovementCommand command) throws Exception
  {
    //System.out.println(sdf.format(new Date())+"->chargeStatusGetPolling..."+command.getOperation());
    getProcessModel().setState(Vehicle.State.CHARGING);
    if(isCommandQueueWithDrawnByCharging()) {
      return;
    
    }
    if((getProcessModel().getEnergyLevel() >= this.vehicle.getEnergyLevelFullyRecharged())){

      ResetChargeDO();
      finishVehicleSimulation(command);
     getProcessModel().setState(Vehicle.State.IDLE);
      cmdListExecEndHook(true);
      return;
    }
    else if(((getProcessModel().getEnergyLevel() >=this.vehicle.getEnergyLevelSufficientlyRecharged())))
    {
 
      finishVehicleSimulation(command);
      getProcessModel().setState(Vehicle.State.IDLE);
       cmdListExecEndHook(true);
       return;
    }

    ((ScheduledExecutorService) getExecutor()).schedule(() -> {
      try {
        chargeStatusGetPolling(command);
      }
      catch (Exception ex) {
        java.util.logging.Logger.getLogger(WECOCommunicationAdapter.class.getName()).log(Level.SEVERE, null, ex);
      }
    },
                                                        2000,
                                                        TimeUnit.MILLISECONDS);
  }

**Comment by swltr** [comment]
@yuantingxue2028 Parking and recharging orders are created automatically by the default dispatcher when configured accordingly - see [the user's guide](https://opentcs.org/docs/6/user/opentcs-users-guide.html#_default_dispatcher_configuration_entries), specifically the configuration entries `defaultdispatcher.parkIdleVehicles` and `defaultdispatcher.rechargeIdleVehicles`.

The transport order to send the vehicle parking/charging is flagged as _dispensable_, which means the default dispatcher will withdraw it automatically in case there is another transport order the vehicle can take. (In other words: you do not have to stop the charging order explicitly in openTCS - the vehicle will automatically get a new transport order once it is in a state in which it can take one.)

**Reply by yuantingxue2028** [reply]
I have tested it, and the official system can achieve the functionality you mentioned, but my rewritten code cannot achieve it. Can you help me see what went wrong in my program? If it's convenient, could you give me an email address so I can send you the program for you to take a look? Thank you very much！！

**Reply by swltr** [reply]
@yuantingxue2028 If you would like someone to take a look at your code, I suggest you post it here or - if it's too much for a post, upload it into a repository here on GitHub and link to it.

I cannot promise I'll have the time to look at it, but maybe someone else does.
